cpp/src/arrow/engine/substrait/expression_internal.cc

```diff
@@ -295,7 +295,8 @@ Result<compute::Expression> FromProto(const substrait::Expression& expr,
 Result<Datum> FromProto(const substrait::Expression::Literal& lit,
                         const ExtensionSet& ext_set,
                         const ConversionOptions& conversion_options) {
-  if (lit.nullable()) {
```

Comment:
`Expression.Literal.{null, empty_list, empty_map}` are all `Type`s, which can independently express nullability. There's [a comment](https://github.com/substrait-io/substrait/blob/f3f6bdc947e689e800279666ff33f118e42d2146/proto/substrait/algebra.proto#L565-L567) for `Expression.Literal.null` which indicates that Literal.nullable should be left false since its type should carry that property. In light of this I'd propose we validate that if `lit` contains one of `Expression.Literal.{null, empty_list, empty_map}` then `lit.nullable()` must be false regardless of the conversion options

Reply:
I don't think that comment says it must be set to false.  It simply says the field does not apply.

I'm not sure I support being pedantic here for no reason as we are not a Substrait validator.  In general, if an option doesn't make sense, but we can safely ignore it, I think we should.

Reply:
Also, for a list, the type-nullability applies to the values in the list and not the lists themselves.  Same goes for maps.  The key/value nullability is distinct from whether you can have an entirely null map.

So this just leaves the null case and it is very weird to expect producers to mark a null literal as non-nullable.